tests/lib/rules/no-redeclare.js

```diff
@@ -716,3 +716,640 @@ ruleTester.run("no-redeclare", rule, {
+						id: "top",
+					},
+					messageId: "redeclaredAsBuiltin",
+					type: "Identifier",
+				},
+			],
+		},
+		{
+			code: `
+	  var a;
+	  var { a = 0, b: Object = 0 } = {};
+			`,
+			options: [{ builtinGlobals: true }],
+			languageOptions: { parserOptions: { ecmaVersion: 6 } },
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+					type: "Identifier",
+				},
+				{
+					data: {
+						id: "Object",
+					},
+					messageId: "redeclaredAsBuiltin",
+					type: "Identifier",
+				},
+			],
+		},
+		{
+			code: `
+	  var a;
+	  var { a = 0, b: Object = 0 } = {};
+			`,
+			options: [{ builtinGlobals: true }],
+			languageOptions: {
+				parserOptions: { ecmaVersion: 6, sourceType: "module" },
+			},
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+					type: "Identifier",
+				},
+			],
+		},
+		{
+			code: `
+	  var a;
+	  var { a = 0, b: Object = 0 } = {};
+			`,
+			options: [{ builtinGlobals: true }],
+			languageOptions: {
+				parserOptions: {
+					ecmaFeatures: { globalReturn: true },
+					ecmaVersion: 6,
+				},
+			},
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+					type: "Identifier",
+				},
+			],
+		},
+		{
+			code: `
+	  var a;
+	  var { a = 0, b: Object = 0 } = {};
+			`,
+			options: [{ builtinGlobals: false }],
+			languageOptions: { parserOptions: { ecmaVersion: 6 } },
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+					type: "Identifier",
+				},
+			],
+		},
+
+		// Notifications of readonly are moved from no-undef: https://github.com/eslint/eslint/issues/4504
+		{
+			code: "/*global b:false*/ var b = 1;",
+			options: [{ builtinGlobals: true }],
+			errors: [
+				{
+					data: {
+						id: "b",
+					},
+					messageId: "redeclaredBySyntax",
+					type: "Block",
+				},
+			],
+		},
+
+		{
+			code: `
+	  type T = 1;
+	  type T = 2;
+			`,
+			errors: [
+				{
+					data: {
+						id: "T",
+					},
+					line: 3,
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  type NodeListOf = 1;
+			`,
+			options: [{ builtinGlobals: true }],
+			languageOptions: {
+				parserOptions: {
+					lib: ["dom"],
+					sourceType: "script",
+				},
+			},
+			errors: [
+				{
+					data: {
+						id: "NodeListOf",
+					},
+					messageId: "redeclaredAsBuiltin",
+				},
+			],
+		},
+		{
+			code: `
+	  function A() {}
+	  class A {}
+			`,
+
+			errors: [
+				{
+					data: {
+						id: "A",
+					},
+					line: 3,
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  function A() {}
+	  class A {}
+	  namespace A {}
+			`,
+			errors: [
+				{
+					data: {
+						id: "A",
+					},
+					line: 3,
+					messageId: "redeclared",
+				},
+				{
+					data: {
+						id: "A",
+					},
+					line: 4,
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  type something = string;
+	  const something = 2;
```

Comment:
I'm not sure about this case. TypeScript considers it valid code since the two declarations don't collide: the first one is a type, the second one is value. `@typescript-eslint/no-redeclare` on the other hand flags this case as invalid because the declarations use the same identifier:

[typescript-eslint Playground](https://typescript-eslint.io/play/#ts=5.8.3&fileType=.ts&code=C4TwDgpgBAzg9gWwsAFgSwHYHMoF5bABOmWA3AFADGcGMwsiy62eUATKUA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1ukQBNEy8AIY90YANrhsORNGgdokKRAC6UgL4h1QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

I'm not sure if the combination of a value and type with the same name should be forbidden.